### DISCUSSION TITLE
When using a Dockerfile for my Dagster jobs, How do I run an additional shell script before my Dagster code loads?

### DISCUSSION THREAD

**Discussion by gibsondan** [dagster-io/dagster#12687] [Q&A] [answered]
When running Dagster in Docker/K8s/ECS, I want to run an additional script before any Dagster code runs (for example, to run a script that adds environment variables). How can I do this?

**Comment by gibsondan** [comment, marked as answer]
Each of the Dagster run launchers overrides the Dockerfile's CMD with a Dagster-specific command (`dagster api execute_run`), but does *not* override the Dockerfile's ENTRYPOINT. This means that you can provide a custom ENTRYPOINT and it will still run before any Dagster code executes. However, it's important that this entrypoint still run the passed-in CMD.

So you can add a shell script to your Dockerfile like this (called `dagster_entrypoint.sh` here):

```
#!/bin/sh

# Add any additional commands here
echo "Running additional command before CMD"

# Run the passed-in Dagster CMD
exec "$@"
```

and then set that shell script as your ENTRYPOINT in your Dockerfile:
```
FROM <your base image>

COPY dagster_entrypoint.sh /dagster_entrypoint.sh
RUN chmod +x /dagster_entrypoint.sh

ENTRYPOINT ["/dagster_entrypoint.sh"]
```

(Kubernetes also provides ways to override the ENTRYPOINT without setting it in your Dockerfile - confusingly, in Kubernetes the ENTRYPOINT is called the "command" and the CMD is called "args". For instructions on how to do this, see https://docs.dagster.io/deployment/guides/kubernetes/customizing-your-deployment)

**Comment by slopp** [comment]
If you are using conda to manage python dependencies, you can use a similar ENTRYPOINT approach to ensure Dagster code runs within an active conda environment. Your Dockerfile might look like:

```
FROM continuumio/miniconda


WORKDIR /opt/dagster/app
COPY ./ ./

RUN conda env create -f environment.yml

# Make RUN commands use the new environment:
SHELL ["conda", "run", "-n", "dagster-ml", "/bin/bash", "-c"]

# pip install packages into the conda env that are not available in conda channels
RUN pip install dagster dagster-cloud

# Code to prefix the run CMD dagster cloud will provide
ENTRYPOINT ["conda", "run", "-n", "dagster-ml"]

# to test
# docker run <image name> dagster --version
# docker run <image name> python -c "import pandas"
```
